# Fiber Maintenance Analytics - Data Exploration

This notebook explores the fiber maintenance dataset to understand its structure, column names, data types, and values before finalizing the project.

## 1. Import Required Libraries

In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
from dotenv import load_dotenv
import os
import warnings
warnings.filterwarnings('ignore')

# Load environment variables
load_dotenv()

print("Libraries imported successfully!")

Libraries imported successfully!


## 2. Load Sample Dataset from Database

In [6]:
# Database connection settings
# Using Configuration 1 (Remote Server) - change to 2 for Local Server
config_number = 1

# Get database credentials from environment
DB_NAME = os.getenv(f'DB_NAME_{config_number}', 'fiber_maintainance_department')
DB_HOST = os.getenv(f'DB_HOST_{config_number}', 'localhost')
DB_USER = os.getenv(f'DB_USER_{config_number}', 'root')
DB_PASSWORD = os.getenv(f'DB_PASSWORD_{config_number}', '')

print(f"Connecting to: {DB_HOST}/{DB_NAME}")

# Create database connection using mysql-connector-python
try:
    engine = create_engine(f'mysql+mysqlconnector://{DB_USER}:{DB_PASSWORD}@{DB_HOST}/{DB_NAME}')
    
    # Load a sample of the data (limit to 1000 rows for exploration)
    query = """
    SELECT * FROM engineered_tickets 
    LIMIT 1000
    """
    df = pd.read_sql(query, engine)
    print(f"Successfully loaded {len(df)} records!")
except Exception as e:
    print(f"Error connecting to database: {e}")
    print("Please check your .env file and database credentials")

Connecting to: 100.83.80.26/fiber_maintainance_department
Successfully loaded 1000 records!


## 3. Inspect Data Structure

In [7]:
# Dataset shape
print("=" * 50)
print("DATASET SHAPE")
print("=" * 50)
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")
print()

# First few rows
print("=" * 50)
print("FIRST 5 ROWS")
print("=" * 50)
df.head()

DATASET SHAPE
Rows: 1,000
Columns: 41

FIRST 5 ROWS


,INC,ACCESS_GRANTED_TIME,ACCESS_PROVIDER,ACCESS_REQUEST_TIME,ACCESS_SITE,CLEARS,DISPATCHER,ENGINEER1,ENGINEER2,ENGINEER3,...,PATROL,CAUSE,CHALLENGE,EXTERNAL_SLA,EXTERNAL_BREACHED_HOURS,EXTERNAL_BREACHED,INHOUSE_SLA,INHOUSE_BREACHED_HOURS,INHOUSE_BREACHED,DURATION
0,INC000010340145,NaT,None,NaT,None,"CORE PROVISIONING DONE, SPLICING DONE IN 6 JOI...",None,ERIC OMONDI,None,None,...,None,CORE PROVISIONING,None,24.0,29.03860,YES,12.0,41.03860,YES,53:02:19
1,INC000010346336,NaT,None,NaT,None,TEAM RESPLICED CORE 10 AT THE ENTRY MH AT NGOM...,None,ERIC OMONDI,None,None,...,None,CORE BREAK,ACCESS DELAY,4.0,1.96361,YES,2.0,3.96361,YES,05:57:49
2,INC000010357554,NaT,None,NaT,None,POWER FAILURE,PASCAL,ERIC OMONDI,None,None,...,None,POWER ISSUE,None,NaN,NaN,NO,NaN,NaN,NO,01:13:36
3,INC000010373107,NaT,None,NaT,None,None,None,None,None,None,...,None,UNCLASSIFIED,None,4.0,21.65310,YES,2.0,23.65310,YES,25:39:11
4,INC000010379033,NaT,None,NaT,None,CLEARS; TEAM SWAPPED CORE 9 TUBE 4 TO TUBE 8 C...,PASCAL,ERIC OMONDI,None,None,...,None,CORE FAILING MIDSPAN,ACCESS DELAY,4.0,28.04890,YES,2.0,30.04890,YES,32:02:56


In [ ]:
# Last few rows
print("LAST 5 ROWS")
print("=" * 50)
df.tail()

In [ ]:
# Detailed info about the dataset
print("DATASET INFO")
print("=" * 50)
df.info()

## 4. Examine Column Names and Data Types

In [8]:
# All column names
print("ALL COLUMN NAMES")
print("=" * 50)
for i, col in enumerate(df.columns, 1):
    print(f"{i:2}. {col}")

print(f"\nTotal: {len(df.columns)} columns")

ALL COLUMN NAMES
 1. INC
 2. ACCESS_GRANTED_TIME
 3. ACCESS_PROVIDER
 4. ACCESS_REQUEST_TIME
 5. ACCESS_SITE
 6. CLEARS
 7. DISPATCHER
 8. ENGINEER1
 9. ENGINEER2
10. ENGINEER3
11. ESCALATED_TIME
12. LATITUDE
13. LINK_DESCRIPTION
14. LONGITUDE
15. MATERIAL_USED
16. REGION
17. ROOT_CAUSE
18. SECURITY_ARRIVAL_TIME
19. SECURITY_REQUEST_TIME
20. SERVICE
21. SOURCE_FILE
22. STATUS
23. UPTIME
24. CAUSE_OF_FAILURE
25. CITY
26. SITE_ID
27. SITE_NAME
28. NOTES
29. SUMMARY
30. ESCALATED_DAY
31. CLUSTER
32. PATROL
33. CAUSE
34. CHALLENGE
35. EXTERNAL_SLA
36. EXTERNAL_BREACHED_HOURS
37. EXTERNAL_BREACHED
38. INHOUSE_SLA
39. INHOUSE_BREACHED_HOURS
40. INHOUSE_BREACHED
41. DURATION

Total: 41 columns


In [9]:
# Column names with data types
print("COLUMNS AND DATA TYPES")
print("=" * 50)
column_info = pd.DataFrame({
    'Column': df.columns,
    'Data Type': df.dtypes.values,
    'Non-Null Count': df.count().values,
    'Null Count': df.isnull().sum().values
})
column_info

COLUMNS AND DATA TYPES


,Column,Data Type,Non-Null Count,Null Count
0,INC,object,1000,0
1,ACCESS_GRANTED_TIME,datetime64[ns],741,259
2,ACCESS_PROVIDER,object,743,257
3,ACCESS_REQUEST_TIME,datetime64[ns],742,258
4,ACCESS_SITE,object,537,463
5,CLEARS,object,783,217
6,DISPATCHER,object,649,351
7,ENGINEER1,object,807,193
8,ENGINEER2,object,0,1000
9,ENGINEER3,object,0,1000


## 5. View Sample Values and Statistics

In [ ]:
# Numerical columns statistics
print("NUMERICAL COLUMNS STATISTICS")
print("=" * 50)
df.describe()

In [10]:
# Key categorical columns - unique values
key_columns = ['CLUSTER', 'REGION', 'SERVICE', 'CAUSE', 'CHALLENGE', 'EXTERNAL_BREACHED', 'ASSIGNED_ENGINEER']

print("KEY CATEGORICAL COLUMNS - UNIQUE VALUES")
print("=" * 50)
for col in key_columns:
    if col in df.columns:
        print(f"\n{col}:")
        print(f"  Unique values: {df[col].nunique()}")
        print(f"  Top 5 values:")
        top_values = df[col].value_counts().head(5)
        for val, count in top_values.items():
            print(f"    - {val}: {count:,} ({count/len(df)*100:.1f}%)")

KEY CATEGORICAL COLUMNS - UNIQUE VALUES

CLUSTER:
  Unique values: 0
  Top 5 values:

REGION:
  Unique values: 2
  Top 5 values:
    - MOMBASA: 721 (72.1%)
    - NAIROBI: 21 (2.1%)

SERVICE:
  Unique values: 11
  Top 5 values:
    - SDC/FTTS: 442 (44.2%)
    - NPS: 203 (20.3%)
    - FTTB: 179 (17.9%)
    - FTTH: 58 (5.8%)
    - UNCLASSIFIED: 39 (3.9%)

CAUSE:
  Unique values: 31
  Top 5 values:
    - CORE BREAK: 104 (10.4%)
    - CABLE DAMAGE: 98 (9.8%)
    - PATCHING ISSUE: 92 (9.2%)
    - CABLE CUT: 90 (9.0%)
    - UNCLASSIFIED: 88 (8.8%)

CHALLENGE:
  Unique values: 15
  Top 5 values:
    - UNCLASSIFIED: 140 (14.0%)
    - ACCESS DELAY: 126 (12.6%)
    - MULTIPLE POF: 22 (2.2%)
    - HEAVY RAINS: 15 (1.5%)
    - TRAFFIC: 9 (0.9%)

EXTERNAL_BREACHED:
  Unique values: 2
  Top 5 values:
    - NO: 683 (68.3%)
    - YES: 317 (31.7%)


## 6. Check for Missing Values

In [11]:
# Missing values analysis
print("MISSING VALUES ANALYSIS")
print("=" * 50)
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Column': missing.index,
    'Missing Count': missing.values,
    'Missing %': missing_pct.values
})

# Show only columns with missing values
missing_with_nulls = missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False)

if len(missing_with_nulls) > 0:
    print(f"Columns with missing values: {len(missing_with_nulls)}")
    print(missing_with_nulls.to_string(index=False))
else:
    print("No missing values found in the dataset!")

MISSING VALUES ANALYSIS
Columns with missing values: 28
                 Column  Missing Count  Missing %
              ENGINEER3           1000      100.0
              ENGINEER2           1000      100.0
                CLUSTER           1000      100.0
                 PATROL           1000      100.0
  SECURITY_REQUEST_TIME            874       87.4
  SECURITY_ARRIVAL_TIME            872       87.2
               LATITUDE            705       70.5
EXTERNAL_BREACHED_HOURS            683       68.3
              CHALLENGE            648       64.8
              LONGITUDE            539       53.9
            ACCESS_SITE            463       46.3
 INHOUSE_BREACHED_HOURS            435       43.5
          MATERIAL_USED            391       39.1
             DISPATCHER            351       35.1
                  NOTES            302       30.2
       CAUSE_OF_FAILURE            262       26.2
    ACCESS_GRANTED_TIME            259       25.9
    ACCESS_REQUEST_TIME            258      

## 7. Validate Data Quality

In [12]:
# Check for duplicates
print("DATA QUALITY CHECKS")
print("=" * 50)

# Duplicate rows
if 'INC' in df.columns:
    duplicate_incs = df['INC'].duplicated().sum()
    print(f"Duplicate INC (ticket) values: {duplicate_incs}")
    
duplicate_rows = df.duplicated().sum()
print(f"Fully duplicate rows: {duplicate_rows}")

# Date range
if 'CREATED_TIME' in df.columns:
    print(f"\nDate Range:")
    print(f"  Earliest: {df['CREATED_TIME'].min()}")
    print(f"  Latest: {df['CREATED_TIME'].max()}")

# SLA Breach distribution
if 'EXTERNAL_BREACHED' in df.columns:
    print(f"\nSLA Breach Distribution:")
    breach_counts = df['EXTERNAL_BREACHED'].value_counts()
    for status, count in breach_counts.items():
        print(f"  {status}: {count:,} ({count/len(df)*100:.1f}%)")

DATA QUALITY CHECKS
Duplicate INC (ticket) values: 0
Fully duplicate rows: 0

SLA Breach Distribution:
  NO: 683 (68.3%)
  YES: 317 (31.7%)


In [14]:
# Duration/MTTR analysis
if 'DURATION' in df.columns:
    print("DURATION (MTTR) ANALYSIS")
    print("=" * 50)
    
    # Check the format of DURATION column
    print(f"Data type: {df['DURATION'].dtype}")
    print(f"\nSample values:")
    print(df['DURATION'].head(10).tolist())
    
    # Convert duration from HH:MM:SS string to hours
    def duration_to_hours(duration_str):
        try:
            if pd.isna(duration_str):
                return None
            parts = str(duration_str).split(':')
            if len(parts) == 3:
                hours = int(parts[0]) + int(parts[1])/60 + int(parts[2])/3600
                return hours
            return None
        except:
            return None
    
    df['DURATION_HOURS'] = df['DURATION'].apply(duration_to_hours)
    
    print(f"\nConverted to hours:")
    print(f"Mean MTTR: {df['DURATION_HOURS'].mean():.2f} hours")
    print(f"Median MTTR: {df['DURATION_HOURS'].median():.2f} hours")
    print(f"Min MTTR: {df['DURATION_HOURS'].min():.2f} hours")
    print(f"Max MTTR: {df['DURATION_HOURS'].max():.2f} hours")
    print(f"Std Dev: {df['DURATION_HOURS'].std():.2f} hours")
    
    # Check for outliers (> 24h SLA)
    over_sla = (df['DURATION_HOURS'] > 24).sum()
    print(f"\nTickets over 24h SLA: {over_sla:,} ({over_sla/len(df)*100:.1f}%)")

DURATION (MTTR) ANALYSIS
Data type: object

Sample values:
['53:02:19', '05:57:49', '01:13:36', '25:39:11', '32:02:56', '03:49:22', '00:54:10', '10:43:21', '04:30:14', '06:39:14']

Converted to hours:
Mean MTTR: 14.65 hours
Median MTTR: 3.73 hours
Min MTTR: 0.06 hours
Max MTTR: 938.13 hours
Std Dev: 48.13 hours

Tickets over 24h SLA: 111 (11.1%)


## 8. Summary

The `engineered_tickets` table contains **41 columns** with fiber maintenance ticket records. Key columns include:

### Identification & Status
- **INC**: Unique ticket identifier
- **STATUS**: Ticket status
- **SOURCE_FILE**: Data source

### Location & Assignment  
- **REGION**: Geographic region (Mombasa, Nairobi)
- **CLUSTER**: Fiber cluster/zone (currently all null)
- **CITY**: City location
- **SITE_ID/SITE_NAME**: Site information
- **DISPATCHER**: Ticket dispatcher
- **ENGINEER1/ENGINEER2/ENGINEER3**: Assigned engineers

### Service & Issue Details
- **SERVICE**: Type of service (SDC/FTTS, NPS, FTTB, FTTH, etc.)
- **CAUSE**: Root cause category (Core Break, Cable Damage, etc.)
- **CHALLENGE**: Challenges faced during resolution
- **ROOT_CAUSE**: Root cause details
- **CAUSE_OF_FAILURE**: Failure cause
- **SUMMARY**: Ticket summary
- **LINK_DESCRIPTION**: Link description

### Time & SLA Metrics
- **ESCALATED_TIME**: When ticket was escalated
- **DURATION**: Time to resolve (HH:MM:SS format)
- **EXTERNAL_SLA**: External SLA target
- **EXTERNAL_BREACHED**: SLA breach indicator (YES/NO) - 31.7% breach rate
- **EXTERNAL_BREACHED_HOURS**: Hours over SLA if breached
- **INHOUSE_SLA/INHOUSE_BREACHED/INHOUSE_BREACHED_HOURS**: Internal SLA metrics

### Additional Data
- **LATITUDE/LONGITUDE**: GPS coordinates (70% missing)
- **ACCESS_REQUEST_TIME/ACCESS_GRANTED_TIME**: Access timing
- **SECURITY_REQUEST_TIME/SECURITY_ARRIVAL_TIME**: Security timing
- **MATERIAL_USED**: Materials used for repair
- **NOTES**: Additional notes
- **PATROL/CLEARS/UPTIME**: Additional metrics

### Key Statistics
- Mean MTTR: **14.65 hours**
- Median MTTR: **3.73 hours**
- SLA Breach Rate: **31.7%** (317 out of 1000 sampled tickets)
- Tickets over 24h: **11.1%**